### PR TITLE
Add support for exogenous variables to forecasting reduction

### LINKED ISSUE #345: Extend reduction from forecasting to regression to handle exogenous data
**Is your feature request related to a problem? Please describe.**
Currently, the reduction meta-estimators only handle univariate data (we raise a not-implemented error [here](https://github.com/alan-turing-institute/sktime/blob/979aed585983fb820ddb0ca8a619d5730dbed4a7/sktime/forecasting/compose/_reduce.py#L67)). We need to extend it to support multivariate time series data. 

**Describe the solution you'd like**
* Use temporal cross-validation iterator to transform multivariate time series into data frame with lagged values and vector with target values. Currently, the transformation happens [here](https://github.com/alan-turing-institute/sktime/blob/979aed585983fb820ddb0ca8a619d5730dbed4a7/sktime/forecasting/compose/_reduce.py#L73). 
* The `x_windows` list needs to be extended to contain the data from all series. 
* The different formatting of the windows for reduction to time series and tabular regression needs to be adapted too to handle multivariate data. 

**Additional context**
Linked to #228

### PR DESCRIPTION
**Pull request by mloning**
#### Reference Issues/PRs
Closes #345 
Related PR: #699, #792 
Contributors: @mloning, @Lovkush-A

#### What does this implement/fix? Explain your changes.
* adds support for exogenous variables to forecasting reduction (lagged values only, no contemporaneous values yet)
* adds more efficient sliding window transforms for tabular/panel output
* implements `MultioutputTimeSeriesRegressionForecaster`
* works towards unifying tabular and time-series reduction 
* adds more extensive unit tests for reduction approaches

#### What's still missing
For discussion, see #792 
* for exogenous variables, only considers lagged values for same time points as target variable, not contemporaneous values
* more consistent handling of X (#787)
* simplification of checks on forecasting horizon 
* factor out transformation and formatting, so that same transform/format function can be used in fit and predict 
* raising an error if last window contains insufficient data, rather than returning nan values

#### Does your contribution introduce a new dependency? If yes, which one?
No.

#### What should a reviewer concentrate their feedback on?
Main changes are in
* `sktime/forecasting/compose/_reduce.py`
* `sktime/forecasting/compose/tests/test_reduce.py`

Other files are mostly formatting/minor changes. 

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.

##### For new estimators
- [x] I've added the estimator to the online documentation.
- [x] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by mloning** [comment]
@Lovkush-A this implements reduction for lagged exogenous values for both tabular and time series regression, would you like to take a look and give some feedback?

**Comment by Lovkush-A** [comment]
@mloning Yes, I am happy to have a look. I have only had a quick glance so far and my first instinct is that there are quite a lot of changes going on, so it will take a while to go through it all.

I don't know how difficult this is, but is it possible to split up commits into several PRs? What would be the pros and cons of splitting things up?  E.g. have one for the new 'make_reduction', and another which deals with renaming things to include tabular in the name, etc.

**Comment by mloning** [comment]
@Lovkush-A thanks - I don't think I can untangle the changes anymore. I agree, would have been easier to review that way.

**Comment by Lovkush-A** [comment]
@mloning Alright - time for me to dig in! Any tips on doing code reviews?

**Comment by fkiraly** [comment]
@lovkush-A, as @mloning says - it would be better if you could separate into smaller PR.
* Try to separate by "semantic blocks" or individual features that are added.
* if possible: identify features that do not depend on each other, make "parallel" PR if possible; note this may not work if there are conditionalities
* make the change log explicit in the header of each PR - explain what is being changed, and where the feature are located.

For "how-to", a quick hacky way to do this is start off dev for new PR and copy files from your PR state (or parts of files).

**Comment by Lovkush-A** [comment]
@fkiraly Thank you for the tips, Franz! Does that mean restructuring a PR is a part of doing a review?

**Comment by mloning** [comment]
No I think @fkiraly thinks that this is your PR, but it's actually mine! 😁

**Comment by fkiraly** [comment]
I'd see this as a special case of "refactoring" (the coding concept), even if it does not end up changing the eventual difference of code.
Worth doing if the PR becomes too "sprawling" and hard to review (many files changed, multiple features packed together, etc).

**Comment by Lovkush-A** [comment]
How about this: as a part of the review process for this PR, I create a whole bunch of new PRs isolating the various changes? It will force me to really understand what is going on, and be able to provide more robust feedback.

**Comment by mloning** [comment]
@Lovkush-A sounds good

**Comment by fkiraly** [comment]
Quick comment: after a short discussion with @mloning, I noticed that I confused this with #699 of @lovkush-A that @mloning had closed and re-factored in this one. I was also unaware of discussions that had happened.

I'd suggest that the two of you figure out how to best proceed, given discussiosn already happening - apologies for overlooking that there were two PR about the same matter and complicating things.

**Comment by Lovkush-A** [comment]
I tried pushing to this branch but I got error message about permissions:

`remote: Permission to alan-turing-institute/sktime.git denied to Lovkush-A.`

File: sktime/forecasting/compose/_reduce.py [hunk 1]
```diff
@@ -2,108 +2,176 @@
...
+    fh : ForecastingHorizon
+        Forecasting horizon for transformed target variable
+    X : pd.DataFrame, optional (default=None)
+        Exogenous series.
+    scitype : str {"tabular-regressor", "time-series-regressor"}, optional
+        Scitype of estimator to use with transformed data.
+        - If "tabular-regressor", returns X as tabular 2d array
+        - If "time-series-regressor", returns X as panel 3d array
+
+    See section 2.1 of this hackmd file for intended design:
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:71]: Remove link to hackmd - if we want to include this discussion, we should add it to the docstring or user guide.

File: sktime/forecasting/compose/_reduce.py [hunk 2]
```diff
@@ -2,108 +2,176 @@
...
+        j = n_timepoints + effective_window_length - k
+        Zt[i:j, :, k] = z
+
+    # Truncate data, selecting only full windows, discarding incomplete ones.
+    Zt = Zt[effective_window_length:-effective_window_length]
+
+    # Return transformed feature and target variables separately. This excludes
+    # contemporaneous values of the exogenous variables. Including them would lead to
+    # unequal-length data, with more time points for exogenous series than the target
+    # series, with is currently not supported.
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:126]:
```suggestion
    # series, which is currently not supported.
```

File: sktime/forecasting/compose/_reduce.py [hunk 3]
```diff
@@ -2,108 +2,176 @@
...
+    yt = Zt[:, 0, window_length + fh]
+    Xt = Zt[:, :, :window_length]
+
+    if scitype == "tabular-regressor":
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:130]: Add comment here to explain tabularisation

File: sktime/forecasting/compose/_reduce.py [hunk 4]
```diff
@@ -187,69 +221,77 @@ def fit(self, y, X=None, fh=None):
...
+        if not self._is_predictable(y_last):
             return self._predict_nan(fh)
 
-        X_last = self._format_windows([last_window])
+        if self._X is None:
+            n_columns = 1
+        else:
+            # X is ignored here, since we currently only look at lagged values for
+            # exogenous variables and not contemporaneous ones.
+            # X = X.to_numpy()
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:255]: remove

File: sktime/forecasting/compose/_reduce.py [hunk 5]
```diff
@@ -264,327 +306,428 @@ def fit(self, y, X=None, fh=None):
...
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:337]: remove

File: sktime/forecasting/compose/_reduce.py [hunk 6]
```diff
@@ -264,327 +306,428 @@ def fit(self, y, X=None, fh=None):
...
+
+        # If we cannot generate a prediction from the available data, return nan.
+        if not self._is_predictable(y_last):
+            return self._predict_nan(fh)
+
+        # Pre-allocate arrays.
+        if X is None:
+            n_columns = 1
+        else:
+            X = X.to_numpy()
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:399]: remove

File: sktime/forecasting/compose/_reduce.py [hunk 7]
```diff
@@ -264,327 +306,428 @@ def fit(self, y, X=None, fh=None):
...
+    References
+    ----------
+    ..[1] Bontempi, Gianluca & Ben Taieb, Souhaib & Le Borgne, Yann-Aël. (
+    2013).
+      Machine Learning Strategies for Time Series Forecasting.
+    """
+    if step_length != 1:
         raise ValueError(
-            f"Unknown strategy, please provide one of {allowed_strategies}."
+            "`step_length` values different from 1 are no longer " "supported."
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:623]:
```suggestion
            "`step_length` values different from 1 are no longer supported."
```

File: sktime/forecasting/compose/_reduce.py [hunk 8]
```diff
@@ -264,327 +306,428 @@ def fit(self, y, X=None, fh=None):
...
+            f" {valid_scitypes}, but found: {scitype}."
+        )
+    return scitype
+
+
+def _infer_scitype(estimator):
+    # We can check if estimator is an instance of scikit-learn's RegressorMixin or
+    # of sktime's BaseRegressor, otherwise we raise an error. Some time-series
+    # regressor also inherit from scikit-learn classes, hence the order in which we
+    # check matters.
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:694]:
```suggestion
    # check matters and we first need to check for BaseRegressor.
```

File: sktime/forecasting/compose/_reduce.py [hunk 9]
```diff
@@ -264,327 +306,428 @@ def fit(self, y, X=None, fh=None):
...
+        "time-series-regressor": {
             "direct": DirectTimeSeriesRegressionForecaster,
             "recursive": RecursiveTimeSeriesRegressionForecaster,
+            "multioutput": MultioutputTimeSeriesRegressionForecaster,
         },
     }
-    # look up and return forecaster class
-    Forecaster = lookup_table.get(scitype).get(strategy)
-    return Forecaster
+    return registry.get(scitype).get(strategy)
```
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:733]:
```suggestion
    return registry[scitype][strategy]
```

File: sktime/forecasting/compose/tests/test_reduce.py [hunk 10]
```diff
@@ -2,103 +2,379 @@
...
+        # For the recursive strategy, we always use the first-step ahead as the
+        # target vector in the regression problem during training, regardless of the
+        # actual forecasting horizon.
+        effective_window_length = window_length
+    else:
+        # For the other strategies, we split the data taking into account the steps
+        # ahead we want to predict.
+        effective_window_length = window_length + max(fh) - 1
+
+    expected = np.mean(y_train[effective_window_length:])
```
Comment (mloning) [inline sktime/forecasting/compose/tests/test_reduce.py:250]: add comment

**Review by mloning — COMMENTED** [review]
Self-review.

File: sktime/forecasting/compose/_reduce.py (hunk 1 above)
Comment (mloning) [reply sktime/forecasting/compose/_reduce.py:71]: FYI @Lovkush-A

**Comment by mloning** [comment]
@Lovkush-A hopefully all tests pass now, let me know if you're happy with this to be merged as it is.

I added a list of the missing bits and pieces in the PR description at the top, but they are best handled in separate PRs I think.

**Comment by Lovkush-A** [comment]
I'm happy for it to be merged. Having future PRs to tidy things up makes sense.

**Pull request merged**